### PR TITLE
Feature/uttmm 43 Front and back CRUD operations with TypeProject

### PR DESCRIPTION
**Pull request by PanasiukAndriy**

### CODE REVIEW

File: Angular/src/app/app.module.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import { BrowserModule } from '@angular/platform-browser';
-import { NgModule, ErrorHandler } from '@angular/core';
+import { BrowserModule } from "@angular/platform-browser";
+import { NgModule, ErrorHandler } from "@angular/core";
```
Comment (yrikturash) [inline Angular/src/app/app.module.ts]: quotes style should be the same:
```suggestion
import { NgModule, ErrorHandler } from '@angular/core';
```

File: Angular/src/app/project/project.component.ts [hunk 2]
```diff
@@ -1,38 +1,56 @@
...
-  onEdit(project){
+  onEdit(project) {
     this.service.populateForm(project);
     this.dialogService.openFormDialog(ProjectFormComponent);
   }
 
+  onCreateTypeProject() {
+    this.typeProjectService.initializeFormGroup();
+    // this.typeProjectService.openFormDialog(TypeProjectFormComponent);
+    //from SHARED
```
Comment (yrikturash) [inline Angular/src/app/project/project.component.ts]: rename

File: Angular/src/app/shared/api-endpoint.constants.ts [hunk 3]
```diff
@@ -7,3 +6,4 @@ export const requestUrl = baseUrl + "request";
 export const categoryUrl = baseUrl + "category";
 export const participUrl = baseUrl + "participation";
 export const paymentUrl = baseUrl + "payment";
+export const typeProjectUrl = `${baseUrl}typeProject/`;
```
Comment (yrikturash) [inline Angular/src/app/shared/api-endpoint.constants.ts]: please, keep consistency or change everything else:
```suggestion
export const typeProjectUrl =baseUrl + "typeProject/";
```

File: Angular/src/app/type-project/type-project.component.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  onSubmit() {
+    if (this.service.form.invalid) {
+      const controls = this.service.form.controls;
+      Object.keys(controls).forEach(controlName =>
+        controls[controlName].markAsTouched()
+      );
+      return;
+    } else if (!this.service.form.get("id").value) {
```
Comment (yrikturash) [inline Angular/src/app/type-project/type-project.component.ts]:
```suggestion
    } 
    if (!this.service.form.get("id").value) {
```

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+                await _uow.SaveChangesAsync();
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"Error when remove ProjectType by {nameof(ProjectType.Id)}: {id}: ", ex);
+            }
+        }
+
+        public async Task UpdateProjectType(ProjectType projectType, int id)
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]:
```suggestion
        public async Task Update(ProjectType projectType, int id)
```

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+{
+    public class ProjectTypeService : IProjectTypeService
+    {
+        private readonly IUnitOfWork _uow;
+        public ProjectTypeService(IUnitOfWork unitOfWork)
+        {
+            _uow = unitOfWork;
+        }
+
+        public async Task AddProjectType(ProjectType projectType)
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]:
```suggestion
        public async Task Add(ProjectType projectType)
```

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            ProjectType Projecttype = new ProjectType { TypeName = projectType.TypeName };
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]: this is unused

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            if (projectType == null)
+            {
+                throw new NullReferenceException($"{nameof(ProjectType)} can`t be null");
+            }
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]: this check should be in controller and it is ArgumentException

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            try
+            {
+                await _uow.Repository<ProjectType>().Add(projectType);
+                await _uow.SaveChangesAsync();
+            }
+            catch (Exception e)
+            {
+                throw new Exception($"Error when trying to add new {nameof(ProjectType)}");
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]: add an exception to inner:
```suggestion
                throw new Exception($"Error when trying to add new {nameof(ProjectType)}", e);
```

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            }
+        }
+
+        public async Task RemoveProjectType(int id)
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]:
```suggestion
        public async Task Remove(int projectTypeId)
```

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                throw new NullReferenceException($"{nameof(ProjectType)} can`t be null");
+            }
+            try
+            {
+                await _uow.Repository<ProjectType>().Add(projectType);
+                await _uow.SaveChangesAsync();
+            }
+            catch (Exception e)
+            {
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]: add data to
ex.Data[nameof(ProjectType)] = projectType;

File: examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            if (projectType == null)
+            {
+                throw new NullReferenceException($"{nameof(ProjectType)} can`t be null");
+            }
+            try
+            {
+                await _uow.Repository<ProjectType>().Add(projectType);
+                await _uow.SaveChangesAsync();
+            }
+            catch (Exception e)
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Services/ProjectTypeService.cs]:
```suggestion
            catch (Exception ex)
```

File: examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        public ProjectTypeController(IProjectTypeService service)
+        {
+            _service = service;
+        }
+
+        private readonly IProjectTypeService _service;
+
+        [HttpPost]
+        [Route("create/{ProjectType}")]
+        public async Task<IActionResult> PostProjectType([FromBody]ProjectType projectType)
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs]: we should not use entities her - but create separate models, but let it be for now

File: examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using MvcWeb.Services;
+using MvcWeb.TheraLang.Entities;
+
+namespace MvcWeb.TheraLang.Controllers
+{
+    [Route("api/ProjectType")]
```
Comment (yrikturash) [inline examples/MvcWeb/TheraLang/Controllers/ProjectTypeController.cs]:
```suggestion
    [Route("api/projectTypes")]
```
but so not forget update link on front as well

**Comment by yrikturash** [comment]
@PanasiukAndriy please fix the build

**Comment by PanasiukAndriy** [comment]
not relevant

**Pull request closed without merging**